walk/walk.go

```diff
@@ -155,31 +154,50 @@ func visit(c *config.Config, cexts []config.Configurer, isBazelIgnored isIgnored
 	c = configure(cexts, knownDirectives, c, rel, f)
 	wc := getWalkConfig(c)
 
-	if wc.isExcluded(rel, ".") {
+	// Don't recurse into excluded directories
+	if wc.isExcluded(rel) {
 		return
 	}
 
+	// Append an additional element to the relParts array to allow swapping the trailing entry
+	// in each iteration of the loop below.
+	entParts := append(relParts, ".")
+
 	var subdirs, regularFiles []string
 	for _, ent := range ents {
-		base := ent.Name()
-		if isBazelIgnored(path.Join(rel, base)) || wc.isExcluded(rel, base) {
+		name := ent.Name()
+		entRel := path.Join(rel, name)
+
+		// Skip blank entry names, .git directories, ignored and excluded.
+		if name == "" || name == ".git" || isBazelIgnored(entRel) || wc.isExcluded(entRel) {
+			continue
+		}
+
+		entParts[len(entParts)-1] = name
+
+		// Skip git-ignored when the gitignore option is enabled.
+		if wc.gitignoreEnabled && wc.isGitIgnored(entParts, ent.IsDir()) {
 			continue
 		}
-		ent := resolveFileInfo(wc, dir, rel, ent)
+
+		ent := resolveFileInfo(wc, dir, entRel, ent)
 		switch {
 		case ent == nil:
 			continue
 		case ent.IsDir():
-			subdirs = append(subdirs, base)
+			subdirs = append(subdirs, name)
 		default:
-			regularFiles = append(regularFiles, base)
+			regularFiles = append(regularFiles, name)
 		}
 	}
 
 	shouldUpdate := updateRels.shouldUpdate(rel, updateParent)
 	for _, sub := range subdirs {
 		if subRel := path.Join(rel, sub); updateRels.shouldVisit(subRel, shouldUpdate) {
-			visit(c, cexts, isBazelIgnored, knownDirectives, updateRels, wf, filepath.Join(dir, sub), subRel, shouldUpdate)
+			// Create a new slice to allow the nested 'visit' call to modify the slice without affecting the parent.
```

Comment:
do we expect the existing elements to get mutated? Or only append new ones? Asking because if `relParts` has capacity for an extra element, then `subParts` will be a new slice but they will share the same backing store, so mutating it's 0..n-1 index elements will also mutate `relParts`

Reply:
The only mutation is of `entParts` which is local to this method, however `entParts = append(relParts, ".")`.

We might have to think about this more when the loop becomes concurrent though.